nav2_lifecycle_manager/src/lifecycle_manager.cpp

```diff
@@ -126,11 +126,20 @@ LifecycleManager::changeStateForNode(const std::string & node_name, std::uint8_t
 }
 
 bool
-LifecycleManager::changeStateForAllNodes(std::uint8_t transition)
+LifecycleManager::changeStateForAllNodes(std::uint8_t transition, bool reverse_order)
 {
-  for (const auto & kv : node_map_) {
-    if (!changeStateForNode(kv.first, transition)) {
-      return false;
+  if (!reverse_order) {
+    for (auto & node_name : node_names_) {
+      if (!changeStateForNode(node_name, transition)) {
+        return false;
+      }
+    }
+  } else {
+    std::vector<std::string>::reverse_iterator rit;
+    for (rit = node_names_.rbegin(); rit != node_names_.rend(); ++rit) {
```

Comment:
I'd suggest factoring out this core code to a function that takes a start and end iterator. Then the changeStateForAllNodes can call it with forward iterators, and you can create another function `changeStateForAllNodesInReverseOrder` that calls it with reverse iterators.

That way you don't duplicate any code, you get rid of the if statement and you don't have to pass a bool into the function, which is an antipattern of sorts.

Reply:
From your description of this alternative, it seems about the same complexity or more to factor it out even more. I think it's best considered in a further refactor of this class since I mentioned that we could replace the `bringUpNode` that's called in the `startup()` with this `changeStateForAllNodes`. I wanted to get feedback first that there isn't a consequence to transitioning all nodes at once before doing that PR though. As of now, this function is only being called in reverse order anyway.

Reply:
> it seems about the same complexity or more to factor it out even more

It's possible that the alternative would be more lines of code, but every function would be simple, understandable without referring to other code, do one thing and one thing only, avoid duplication of logic, and provide a building block if we ever wanted to transition only a subset of nodes.

I did approve it though. It's your call.